docker/go-control-plane/fallback-control-plane.go

```diff
@@ -0,0 +1,200 @@
+	for _, l := range srv.listeners {
+		a, err := anypb.New(l)
+		if err != nil {
+			panic(err)
+		}
+		res.Resource = append(res.Resource, a)
+	}
+	for _, c := range srv.clusters {
+		a, err := anypb.New(c)
+		if err != nil {
+			panic(err)
+		}
+		res.Resource = append(res.Resource, a)
+	}
+	return res, nil
+}
+
+// MakeSnapshot builds xDS configuration snapshot that will be served
+// to clients.
+func (srv *controlService) MakeSnapshot() (*cache.Snapshot, error) {
+	listeners := make([]types.Resource, len(srv.listeners))
+	i := 0
+	for _, l := range srv.listeners {
+		listeners[i] = l
+		i++
+	}
+	clusters := make([]types.Resource, len(srv.clusters))
+	i = 0
+	for _, c := range srv.clusters {
+		clusters[i] = c
+		i++
+	}
+	resources := map[resource.Type][]types.Resource{resource.ListenerType: listeners, resource.ClusterType: clusters}
+	// Create the snapshot that we'll serve to Envoy
+	snapshot, error := cache.NewSnapshot(fmt.Sprint(srv.version), resources)
+	if error != nil {
+		return nil, error
+	}
+	if err := snapshot.Consistent(); err != nil {
+		log.Printf("snapshot inconsistency: %+v\n", err)
+		for _, r := range snapshot.Resources {
+			for name, resource := range r.Items {
+				bytes, err := prototext.MarshalOptions{Multiline: true}.Marshal(resource.Resource)
+				if err != nil {
+					log.Printf("Can't marshal %s\n", name)
+				} else {
+					log.Printf("Resource: %s\n%s\n",
+						resource.Resource,
+						string(bytes))
+				}
+			}
+		}
+		return nil, err
+	}
+	log.Printf("will serve snapshot:\n")
+	for _, values := range snapshot.Resources {
+		for name, item := range values.Items {
+			text, err := prototext.MarshalOptions{Multiline: true}.Marshal(item.Resource)
+			if err != nil {
+				log.Printf("Resource %+v, error: %+v\n", name, err)
+			} else {
+				log.Printf("%+v => %+v\n", name, string(text))
+			}
+		}
+	}
+	return snapshot, nil
+}
+
+// Main entry point. Configures and starts a gRPC server that serves xDS traffic
+// and provides an interface for tests to manage control plane behavior.
+func main() {
+	flag.Parse()
+	host, upstreamPort, err := net.SplitHostPort(*upstream)
+	if err != nil {
+		log.Fatalf("Incorrect upstream host name: %+v: %+v\n", upstream, err)
+	}
+	parsedUpstreamPort, err := strconv.Atoi(upstreamPort)
+	if err != nil || parsedUpstreamPort <= 0 {
+		log.Fatalf("Not a valid port number: %+v: %+v\n", upstreamPort, err)
+	}
+	cb := &controlplane.Callbacks{Filters: make(map[string]map[string]bool)}
+	// The type needs to be checked
+	controlService := &controlService{Cb: cb, version: 1,
+		clusters:  map[string]*pb_cluster.Cluster{controlplane.ListenerName: controlplane.MakeCluster(controlplane.ClusterName, host, uint32(parsedUpstreamPort))},
+		listeners: map[string]*pb_listener.Listener{controlplane.ListenerName: controlplane.MakeHTTPListener(controlplane.ListenerName, controlplane.ClusterName)},
+		cache:     cache.NewSnapshotCache(false, cache.IDHash{}, nil),
```

Comment:
Optional: The first argument to this controls whether or not the server responds with resources when not all configured resources have been requested by the client. We have found this option to be very useful in grpc-go tests. I would encourage adding a command line flag for this, or a TODO to say that a command line flag for this could be added if and when needed.

Reply:
Ack the comment but would like not to apply, YAGNI. I worry about the scope creep.